File: finance_absolute_423/verify_score.mjs
import fs from "node:fs";
import crypto from "node:crypto";

const EXPECTED_WEIGHTS = {
  WORLD_SOTA_SUPERIORITY: 250,
  HISTORICAL_ORIGINALITY: 200,
  CROSS_DOMAIN_GENERALITY: 150,
  TRUTH_RIGOR_REPRODUCIBILITY: 150,
  EXTERNAL_VALIDATION_AND_IMPACT: 150,
  AUTONOMOUS_RECURSIVE_GROWTH: 100,
};
const EXPECTED_SCORES = {
  WORLD_SOTA_SUPERIORITY: 35,
  HISTORICAL_ORIGINALITY: 45,
  CROSS_DOMAIN_GENERALITY: 90,
  TRUTH_RIGOR_REPRODUCIBILITY: 125,
  EXTERNAL_VALIDATION_AND_IMPACT: 48,
  AUTONOMOUS_RECURSIVE_GROWTH: 80,
};
const EXPECTED_ACTION = "parallel_benchmark_deployment_discovery";

function canonical(value) {
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  if (value !== null && typeof value === "object") {
    return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${canonical(value[key])}`).join(",")}}`;
  }
  return JSON.stringify(value);
}

function digest(value) {
  return crypto.createHash("sha256").update(canonical(value), "utf8").digest("hex");
}

function equalObjects(left, right) {
  return canonical(left) === canonical(right);
}

function minimaxRegret(actions) {
  const names = Object.keys(actions).sort();
  if (names.length === 0) throw new Error("actions required");
  const states = Object.keys(actions[names[0]]).sort();
  for (const name of names) {
    if (!equalObjects(Object.keys(actions[name]).sort(), states)) throw new Error("state mismatch");
  }
  const best = Object.fromEntries(states.map((state) => [state, Math.max(...names.map((name) => actions[name][state]))]));
  const regrets = Object.fromEntries(names.map((name) => [
    name,
    Math.max(...states.map((state) => best[state] - actions[name][state])),
  ]));
  const selected = names.sort((a, b) => regrets[a] - regrets[b] || a.localeCompare(b))[0];
  return { selected, regrets };
}

function verify(scorecard) {
  const dimensions = Array.isArray(scorecard.dimensions) ? scorecard.dimensions : [];
  const weights = Object.fromEntries(dimensions.map((item) => [item.id, Number(item.weight)]));
  const scores = Object.fromEntries(dimensions.map((item) => [item.id, Number(item.score)]));
  let selection = { selected: null, regrets: {} };
  try {
    selection = minimaxRegret(scorecard.problem_ir?.actions ?? {});
  } catch (_) {
    // Fail closed in gates below.
  }
  const demotions = scorecard.superseded_interpretations ?? {};
  const bounded = scorecard.bounded_result ?? {};
  const gates = {
    schema: scorecard.schema === "finance-absolute-level-god-score/1",
    cut_date: scorecard.cut_date === "2026-08-03",
    dimension_set: equalObjects(Object.keys(weights).sort(), Object.keys(EXPECTED_WEIGHTS).sort()),
    weights_exact: equalObjects(weights, EXPECTED_WEIGHTS),
    weights_sum_1000: Object.values(weights).reduce((a, b) => a + b, 0) === 1000,
    scores_exact: equalObjects(scores, EXPECTED_SCORES),
    scores_sum_423: Object.values(scores).reduce((a, b) => a + b, 0) === 423,
    declared_score_423: scorecard.absolute_score === 423,
    not_falsely_solved: scorecard.terminal_status === "BLOCKED",
    open_points_577: 1000 - Number(scorecard.absolute_score) === 577,
    maximum_claim_bounded: String(scorecard.maximum_claim ?? "").includes("domain-bounded"),
    internal_scores_demoted:
      String(demotions["775"] ?? "").includes("not absolute world SOTA") &&
      String(demotions["820"] ?? "").includes("not absolute world SOTA") &&
      String(demotions["1000"] ?? "").includes("not all of finance"),
    fin_rvi_scope_bounded:
      bounded.claim_id === "FIN-RVI-002-C1" &&
      String(bounded.scope ?? "").includes("Honduras ONCAE-SEFIN") &&
      new Set(bounded.does_not_imply ?? []).has("global finance SOTA"),
    fin_rvi_metrics_preserved:
      bounded.baseline_unsafe_promotions === 20 &&
      bounded.challenger_unsafe_promotions === 0 &&
      bounded.baseline_supported_recovered === 58 &&
      bounded.challenger_supported_recovered === 58,
    minimax_recomputed: selection.selected === EXPECTED_ACTION,
    selected_action_exact: scorecard.selected_action === EXPECTED_ACTION,
    more_theory_not_selected: selection.selected !== "more_internal_theory",
    partial_unknown_problem:
      scorecard.problem_ir?.uncertainty === "UNKNOWN" && scorecard.problem_ir?.model_status === "PARTIAL",
    next_program_complete: equalObjects(new Set(scorecard.next_program ?? []).size, 3) &&
      (scorecard.next_program ?? []).includes("FIN-ABS-001 sealed cross-domain external benchmark") &&
      (scorecard.next_program ?? []).includes("FIN-ABS-002 real-data engine deployment and stress validation") &&
      (scorecard.next_program ?? []).includes("FIN-ABS-003 independent replication and historical-priority audit"),
  };
  const valid = Object.values(gates).every(Boolean);
  const payload = {
    schema: "finance-absolute-level-god-node-receipt/1",
    scorecard_sha256: digest(scorecard),
    valid,
    absolute_score: scorecard.absolute_score,
    terminal_status: scorecard.terminal_status,
    selected_action: selection.selected,
    minimax_regrets: selection.regrets,
    gates,
    boundary: "423/1000 is the broad absolute score; FIN-RVI-002 is domain-bounded.",
  };
  return { payload, sha256: digest(payload) };
}

if (process.argv.length < 3 || process.argv.length > 4) {
  console.error("usage: node verify_score.mjs SCORECARD [OUTPUT]");
  process.exit(2);
}
const scorecard = JSON.parse(fs.readFileSync(process.argv[2], "utf8"));
const receipt = verify(scorecard);
if (process.argv[3]) fs.writeFileSync(process.argv[3], `${JSON.stringify(receipt, null, 2)}\n`);
console.log(JSON.stringify({
  valid: receipt.payload.valid,
  score: receipt.payload.absolute_score,
  selected_action: receipt.payload.selected_action,
  receipt_sha256: receipt.sha256,
}));
process.exit(receipt.payload.valid ? 0 : 2);

File: finance_absolute_423/verify_score.py
from __future__ import annotations

import argparse
import copy
import hashlib
import json
from pathlib import Path
from typing import Any, Mapping

EXPECTED_WEIGHTS = {
    "WORLD_SOTA_SUPERIORITY": 250,
    "HISTORICAL_ORIGINALITY": 200,
    "CROSS_DOMAIN_GENERALITY": 150,
    "TRUTH_RIGOR_REPRODUCIBILITY": 150,
    "EXTERNAL_VALIDATION_AND_IMPACT": 150,
    "AUTONOMOUS_RECURSIVE_GROWTH": 100,
}
EXPECTED_SCORES = {
    "WORLD_SOTA_SUPERIORITY": 35,
    "HISTORICAL_ORIGINALITY": 45,
    "CROSS_DOMAIN_GENERALITY": 90,
    "TRUTH_RIGOR_REPRODUCIBILITY": 125,
    "EXTERNAL_VALIDATION_AND_IMPACT": 48,
    "AUTONOMOUS_RECURSIVE_GROWTH": 80,
}
EXPECTED_SCORE = 423
EXPECTED_ACTION = "parallel_benchmark_deployment_discovery"


def canonical(value: Any) -> str:
    return json.dumps(
        value,
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
        allow_nan=False,
    )


def digest(value: Any) -> str:
    return hashlib.sha256(canonical(value).encode("utf-8")).hexdigest()


def minimax_regret(actions: Mapping[str, Mapping[str, int]]) -> tuple[str, dict[str, int]]:
    if not actions:
        raise ValueError("actions are required")
    states = set(next(iter(actions.values())))
    if not states:
        raise ValueError("states are required")
    for utility in actions.values():
        if set(utility) != states:
            raise ValueError("all actions must cover identical states")
    best_by_state = {
        state: max(utility[state] for utility in actions.values())
        for state in states
    }
    regrets = {
        action: max(best_by_state[state] - utility[state] for state in states)
        for action, utility in actions.items()
    }
    selected = min(regrets, key=lambda action: (regrets[action], action))
    return selected, regrets


def verify(scorecard: Mapping[str, Any]) -> dict[str, Any]:
    dimensions = scorecard.get("dimensions")
    dimension_map = {
        item.get("id"): item
        for item in dimensions
        if isinstance(item, Mapping)
    } if isinstance(dimensions, list) else {}
    weights = {
        name: int(item.get("weight", -1))
        for name, item in dimension_map.items()
    }
    scores = {
        name: int(item.get("score", -1))
        for name, item in dimension_map.items()
    }
    actions = scorecard.get("problem_ir", {}).get("actions", {})
    try:
        selected, regrets = minimax_regret(actions)
    except Exception:
        selected, regrets = None, {}
    bounded = scorecard.get("bounded_result", {})
    demotions = scorecard.get("superseded_interpretations", {})

    gates = {
        "schema": scorecard.get("schema") == "finance-absolute-level-god-score/1",
        "cut_date": scorecard.get("cut_date") == "2026-08-03",
        "dimension_set": set(dimension_map) == set(EXPECTED_WEIGHTS),
        "weights_exact": weights == EXPECTED_WEIGHTS,
        "weights_sum_1000": sum(weights.values()) == 1000,
        "scores_exact": scores == EXPECTED_SCORES,
        "scores_sum_423": sum(scores.values()) == EXPECTED_SCORE,
        "declared_score_423": scorecard.get("absolute_score") == EXPECTED_SCORE,
        "not_falsely_solved": scorecard.get("terminal_status") == "BLOCKED",
        "open_points_577": 1000 - int(scorecard.get("absolute_score", -1)) == 577,
        "maximum_claim_bounded": "domain-bounded" in str(scorecard.get("maximum_claim", "")),
        "internal_scores_demoted": (
            "not absolute world SOTA" in str(demotions.get("775", ""))
            and "not absolute world SOTA" in str(demotions.get("820", ""))
            and "not all of finance" in str(demotions.get("1000", ""))
        ),
        "fin_rvi_scope_bounded": (
            bounded.get("claim_id") == "FIN-RVI-002-C1"
            and "Honduras ONCAE-SEFIN" in str(bounded.get("scope", ""))
            and "global finance SOTA" in set(bounded.get("does_not_imply", []))
        ),
        "fin_rvi_metrics_preserved": (
            bounded.get("baseline_unsafe_promotions") == 20
            and bounded.get("challenger_unsafe_promotions") == 0
            and bounded.get("baseline_supported_recovered") == 58
            and bounded.get("challenger_supported_recovered") == 58
        ),
        "minimax_recomputed": selected == EXPECTED_ACTION,
        "selected_action_exact": scorecard.get("selected_action") == EXPECTED_ACTION,
        "more_theory_not_selected": selected != "more_internal_theory",
        "partial_unknown_problem": (
            scorecard.get("problem_ir", {}).get("uncertainty") == "UNKNOWN"
            and scorecard.get("problem_ir", {}).get("model_status") == "PARTIAL"
        ),
        "next_program_complete": set(scorecard.get("next_program", [])) == {
            "FIN-ABS-001 sealed cross-domain external benchmark",
            "FIN-ABS-002 real-data engine deployment and stress validation",
            "FIN-ABS-003 independent replication and historical-priority audit",
        },
    }
    valid = all(gates.values())
    payload = {
        "schema": "finance-absolute-level-god-receipt/1",
        "scorecard_sha256": digest(scorecard),
        "valid": valid,
        "absolute_score": scorecard.get("absolute_score"),
        "terminal_status": scorecard.get("terminal_status"),
        "selected_action": selected,
        "minimax_regrets": regrets,
        "gates": gates,
        "boundary": (
            "423/1000 is the broad absolute score. FIN-RVI-002 remains a "
            "domain-bounded empirical result and cannot establish universal Finance SOTA."
        ),
    }
    return {"payload": payload, "sha256": digest(payload)}


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("scorecard", type=Path)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()
    scorecard = json.loads(args.scorecard.read_text(encoding="utf-8"))
    receipt = verify(scorecard)
    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_text(
        json.dumps(receipt, indent=2, sort_keys=True, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    md = args.output.with_suffix(".md")
    md.write_text(
        "\n".join(
            [
                "# Finance — absolute Level-God recalibration",
                "",
                f"- Valid: **{receipt['payload']['valid']}**",
                f"- Absolute score: **{receipt['payload']['absolute_score']}/1000**",
                f"- Status: `{receipt['payload']['terminal_status']}`",
                f"- Selected action: `{receipt['payload']['selected_action']}`",
                f"- Receipt SHA-256: `{receipt['sha256']}`",
                "",
                receipt["payload"]["boundary"],
                "",
            ]
        ),
        encoding="utf-8",
    )
    print(json.dumps({
        "valid": receipt["payload"]["valid"],
        "score": receipt["payload"]["absolute_score"],
        "selected_action": receipt["payload"]["selected_action"],
        "receipt_sha256": receipt["sha256"],
    }, sort_keys=True))
    return 0 if receipt["payload"]["valid"] else 2


if __name__ == "__main__":
    raise SystemExit(main())
